### PR TITLE
[Merged by Bors] - feat: Define Hamiltonian paths and cycles

### PR DESCRIPTION
**Pull request by rishicomplex**
Co-authored-by: Bhavik Mehta <bhavikmehta8@gmail.com>
Co-authored-by: Linus Sommer
Co-authored-by: Lode Vermeulen

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+import Mathlib.Combinatorics.SimpleGraph.Connectivity
+import Mathlib.Combinatorics.SimpleGraph.Trails
+import Mathlib.Algebra.BigOperators.Basic
+
+/-!
+# Hamiltonian Graphs
+
+In this file we introduce `Hamitonian paths`, `cylces` and `graphs`,
+two main concepts in the theory of graphs.
```
[the remark this answers is not in the document]
Comment (linus-md) [reply Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean]: Ooops

**Comment by b-mehta** [comment]
@linus-md We definitely don't want the Dirac file to be in the branch!

**Comment by linus-md** [comment]
@b-mehta I See, I was wondering what happened.

**Comment by YaelDillies** [comment]
I just pushed a few sorries for you to prove. Do you mind trying them?

**Comment by YaelDillies** [comment]
Yeah, I cleaned up already once, so the sorries are the most blocking part. Please also merge master and ensure CI passes.

This PR was written during the Bonn workshop last September and I believe the original authors have gone to do other things, so feel free to take over!

**Comment by LodeVermeulen** [comment]
It seems I don't have push access to this branch. Could that be right? Could I get the access rights?

**Comment by kmill** [comment]
It would be good to coordinate with @linus-md, @rishicomplex, and @b-mehta to be sure that they would appreciate this PR being dusted off and continued (I assume they would, but it's more polite to check rather than assume!)

For getting push access @LodeVermeulen, please go to Zulip [and request write access](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/github.20permission).

**Comment by linus-md** [comment]
I would be very happy if this gets finished! @LodeVermeulen @kmill

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean [hunk 2]
```diff
@@ -1016,8 +1015,6 @@ theorem isCycle_copy {u u'} (p : G.Walk u u) (hu : u = u') :
   rfl
 #align simple_graph.walk.is_cycle_copy SimpleGraph.Walk.isCycle_copy
 
-lemma IsCycle.not_Nil {p : G.Walk v v} (hp : IsCycle p) : ¬ p.Nil := (hp.ne_nil ·.eq_nil)
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean]: You shouldn't drop this lemma. It's generally useful

Comment (LodeVermeulen) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity.lean]: Oops, that was an accident

Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity.lean]: While you're at it, it should be called `IsCycle.not_nil`, not `IsCycle.not_Nil`.

File: Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean [hunk 3]
```diff
@@ -130,7 +132,12 @@ lemma IsHamiltonian.connected (hG : G.IsHamiltonian) : G.Connected where
     · rfl
     have : Nontrivial α := ⟨a, b, hab⟩
     obtain ⟨c, p, hp⟩ := hG Fintype.one_lt_card.ne'
-    sorry
+    have a_mem := Walk.IsHamiltonianCycle.mem_support p hp a
+    have b_mem := Walk.IsHamiltonianCycle.mem_support p hp b
+    have walk_a := p.takeUntil a a_mem
+    have walk_b := p.takeUntil b b_mem
+    have walk_a_b := (id walk_a.reverse).append walk_b
+    exact Walk.reachable walk_a_b
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean:140]:
```suggestion
    exact ((p.takeUntil a a_mem).reverse.append $ p.takeUntil b b_mem).reachable 
```

File: Mathlib/Algebra/BigOperators/Basic.lean [hunk 4]
```diff
@@ -2463,6 +2463,11 @@ theorem prod_toFinset {M : Type*} [DecidableEq α] [CommMonoid M] (f : α → M)
 #align list.prod_to_finset List.prod_toFinset
 #align list.sum_to_finset List.sum_toFinset
 
+@[simp]
+theorem toFinset_sum_count_eq [DecidableEq α] (s : List α) :
+    ∑ a in s.toFinset, s.count a = s.length :=
+  by simpa using (Finset.sum_list_map_count s (fun _ => (1 : ℕ))).symm
```
Comment (YaelDillies) [inline Mathlib/Algebra/BigOperators/Basic.lean:2469]:
```suggestion
theorem sum_toFinset_count_eq [DecidableEq α] (l : List α) :
    ∑ a in l.toFinset, m.count a = l.length :=
  by simpa using (Finset.sum_list_map_count m fun _ => (1 : ℕ)).symm
```

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean [hunk 5]
```diff
@@ -877,6 +877,12 @@ lemma not_nil_iff {p : G.Walk v w} :
     ¬ p.Nil ↔ ∃ (u : V) (h : G.Adj v u) (q : G.Walk u w), p = cons h q := by
   cases p <;> simp [*]
 
+/-- A walk with its endpoints defeq is `Nil` if and only if it is equal to `nil`. -/
+lemma Nil_iff_eq_nil : ∀ {p : G.Walk v v}, p.Nil ↔ p = nil
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:881]:
```suggestion
@[simp]
lemma nil_iff_eq_nil : ∀ {p : G.Walk v v}, p.Nil ↔ p = nil
```

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean [hunk 6]
```diff
@@ -877,6 +877,12 @@ lemma not_nil_iff {p : G.Walk v w} :
...
+  | .nil | .cons _ _ => by simp
+
+lemma Nil.eq_nil {p : G.Walk v v} (hp : p.Nil) : p = Walk.nil := Nil_iff_eq_nil.1 hp
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:884]:
```suggestion
alias ⟨Nil.eq_nil, _⟩ := nil_iff_eq_nil
```

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean [hunk 7]
```diff
@@ -925,6 +931,10 @@ variable {x y : V} -- TODO: rename to u, v, w instead?
     (p.copy hx hy).Nil = p.Nil := by
   subst_vars; rfl
 
+lemma support_tail {p : G.Walk v v} (hp : ¬p.Nil) :
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:934]:
```suggestion
@[simp] lemma support_tail (p : G.Walk v v) (hp : ¬p.Nil) :
```

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean [hunk 8]
```diff
@@ -994,6 +1004,8 @@ theorem isCircuit_copy {u u'} (p : G.Walk u u) (hu : u = u') :
   rfl
 #align simple_graph.walk.is_circuit_copy SimpleGraph.Walk.isCircuit_copy
 
+lemma IsCircuit.not_Nil {p : G.Walk v v} (hp : IsCircuit p) : ¬ p.Nil := (hp.ne_nil ·.eq_nil)
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:1007]:
```suggestion
lemma IsCircuit.not_nil {p : G.Walk v v} (hp : IsCircuit p) : ¬ p.Nil := (hp.ne_nil ·.eq_nil)
```

File: Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+lemma IsHamiltonianCycle_def {p : G.Walk a a} :
+    p.IsHamiltonianCycle ↔ ∃ h : p.IsCycle, (p.tail h.not_nil).IsHamiltonian :=
+  ⟨fun ⟨h, h'⟩ ↦ ⟨h, h'⟩, fun ⟨h, h'⟩ ↦ ⟨h, h'⟩⟩
+
+lemma IsHamiltonianCycle_iff {p : G.Walk a a} :
+    p.IsHamiltonianCycle ↔ p.IsCycle ∧ ∀ a, (support p).tail.count a = 1 := by
+  simp only [IsHamiltonianCycle_def, IsHamiltonian, support_tail, exists_prop]
+
+/-- A hamiltonian cycle visits every vertex. -/
+lemma IsHamiltonianCycle.mem_support (p : G.Walk a a) (hp : p.IsHamiltonianCycle) (b : α) :
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean:98]:
```suggestion
lemma IsHamiltonianCycle.mem_support (hp : p.IsHamiltonianCycle) (b : α) :
```

File: Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    · cases hp.ne_nil rfl
+    simp only [support_cons, List.count_cons, List.map_cons, List.head_cons, hf.injective.eq_iff,
+      add_tsub_cancel_right]
+    exact hp.isHamiltonian_tail _
+
+lemma IsHamiltonianCycle_def {p : G.Walk a a} :
+    p.IsHamiltonianCycle ↔ ∃ h : p.IsCycle, (p.tail h.not_nil).IsHamiltonian :=
+  ⟨fun ⟨h, h'⟩ ↦ ⟨h, h'⟩, fun ⟨h, h'⟩ ↦ ⟨h, h'⟩⟩
+
+lemma IsHamiltonianCycle_iff {p : G.Walk a a} :
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean:93]:
```suggestion
lemma IsHamiltonianCycle_iff :
```

File: Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+  isHamiltonian_tail := by
+    simp [IsHamiltonian, support_tail, hf.surjective.forall, List.count_tail, hf.injective]
+    intro x
+    rcases p with (_ | ⟨y, p⟩)
+    · cases hp.ne_nil rfl
+    simp only [support_cons, List.count_cons, List.map_cons, List.head_cons, hf.injective.eq_iff,
+      add_tsub_cancel_right]
+    exact hp.isHamiltonian_tail _
+
+lemma IsHamiltonianCycle_def {p : G.Walk a a} :
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean:89]:
```suggestion
lemma IsHamiltonianCycle_def :
```

File: Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+  rw [← length_tail_add_one hp.not_nil, hp.isHamiltonian_tail.length_eq, Nat.sub_add_cancel]
+  rw [Nat.succ_le, Fintype.card_pos_iff]
+  exact ⟨a⟩
+
+lemma IsHamiltonianCycle.count_support_self {p : G.Walk a a} (hp : p.IsHamiltonianCycle) :
+    p.support.count a = 2 := by
+  rw [support_eq_cons, List.count_cons_self, ← support_tail, hp.isHamiltonian_tail]
+
+lemma IsHamiltonianCycle.support_count_of_ne {p : G.Walk a a} (hp : p.IsHamiltonianCycle)
+    (h : a ≠ b) : p.support.count b = 1 := by
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean:113]:
```suggestion
lemma IsHamiltonianCycle.support_count_of_ne (hp : p.IsHamiltonianCycle) (h : a ≠ b) :
    p.support.count b = 1 := by
```

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean [hunk 13]
```diff
@@ -925,6 +931,10 @@ variable {x y : V} -- TODO: rename to u, v, w instead?
     (p.copy hx hy).Nil = p.Nil := by
   subst_vars; rfl
 
+@[simp] lemma support_tail (p : G.Walk v v) (hp : ¬p.Nil) :
```
[the remark this answers is not in the document]
Comment (LodeVermeulen) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:934]: I didn't know you could do that, neat :).

[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:934]: The same way as you could write `example (a) : a + 1 = 1 + a := sorry`.

However you seem to have removed it from `cons_support_tail`, not `support_tail`!

File: Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+
+lemma IsHamiltonianCycle.isCycle (hp : p.IsHamiltonianCycle) : p.IsCycle :=
+  hp.toIsCycle
+
+lemma IsHamiltonianCycle.map {H : SimpleGraph β} (f : G →g H) (hf : Bijective f)
+    (hp : p.IsHamiltonianCycle) : (p.map f).IsHamiltonianCycle where
+  toIsCycle := by
+    rw [map_isCycle_iff_of_injective]
+    exact isCycle hp
+    exact Bijective.injective hf
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean:78]:
```suggestion
  toIsCycle := hp.isCycle.map hf.injective
```
once you have rename `map_isCycle_of_injective` to `IsCycle.map`

File: Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+      add_tsub_cancel_right]
+    exact hp.isHamiltonian_tail _
+
+lemma IsHamiltonianCycle_def :
+    p.IsHamiltonianCycle ↔ ∃ h : p.IsCycle, (p.tail h.not_nil).IsHamiltonian :=
+  ⟨fun ⟨h, h'⟩ ↦ ⟨h, h'⟩, fun ⟨h, h'⟩ ↦ ⟨h, h'⟩⟩
+
+lemma IsHamiltonianCycle_iff :
+    p.IsHamiltonianCycle ↔ p.IsCycle ∧ ∀ a, (support p).tail.count a = 1 := by
+  simp only [IsHamiltonianCycle_def, IsHamiltonian, support_tail, exists_prop]
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean:94]: Not super clear which one should be the `_def`. Maybe rename to `isHamiltonianCycle_isCycle_and_isHamiltonian_tail` and `isHamiltonianCycle_iff_isCycle_and_support_count_tail_eq_one`?

Also note that `IsHamiltonianCycle_something` is wrong. It must be in `lowerCamelCase`. Compare this to `IsHamiltonianCycle.something`, where it is in `UpperCamelCase` (and `something` is in `lowerCamelCase`).

File: Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean (hunk 14 above)
Comment (LodeVermeulen) [reply Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean]: Done. I could not get it to look exactly like this, but I think I got it close enough.

File: Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean (hunk 15 above)
Comment (LodeVermeulen) [reply Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean]: I like these names a lot better as well. Done.

File: Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean [hunk 16]
```diff
@@ -72,10 +72,7 @@ lemma IsHamiltonianCycle.isCycle (hp : p.IsHamiltonianCycle) : p.IsCycle :=
 
 lemma IsHamiltonianCycle.map {H : SimpleGraph β} (f : G →g H) (hf : Bijective f)
     (hp : p.IsHamiltonianCycle) : (p.map f).IsHamiltonianCycle where
-  toIsCycle := by
-    rw [map_isCycle_iff_of_injective]
-    exact isCycle hp
-    exact Bijective.injective hf
+  toIsCycle := isCycle.map hf.injective hp.isCycle
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean:75]:
```suggestion
  toIsCycle := hp.isCycle.map hf.injective 
```
once you fix the name

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean [hunk 17]
```diff
@@ -1724,8 +1724,8 @@ theorem map_isCycle_iff_of_injective {p : G.Walk u u} (hinj : Function.Injective
     support_map, ← List.map_tail, List.nodup_map_iff hinj]
 #align simple_graph.walk.map_is_cycle_iff_of_injective SimpleGraph.Walk.map_isCycle_iff_of_injective
 
-alias ⟨_, map_isCycle_of_injective⟩ := map_isCycle_iff_of_injective
-#align simple_graph.walk.map_is_cycle_of_injective SimpleGraph.Walk.map_isCycle_of_injective
+alias ⟨_, isCycle.map⟩ := map_isCycle_iff_of_injective
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:1727]:
```suggestion
alias ⟨_, IsCycle.map⟩ := map_isCycle_iff_of_injective
```

File: Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean (hunk 16 above)
Comment (LodeVermeulen) [reply Mathlib/Combinatorics/SimpleGraph/Hamiltonian.lean:75]: Ah, I see what is going on now, thanks :).

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean (hunk 13 above)
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:934]: > However you seem to have removed it from `cons_support_tail`, not `support_tail`!

This still applies!

**Comment by YaelDillies** [comment]
You seem to have missed https://github.com/leanprover-community/mathlib4/pull/7102#discussion_r1630267822

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean (hunk 13 above)
[the remark this answers is not in the document]
Comment (LodeVermeulen) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity.lean]: My bad! Changed it now.

**Review by YaelDillies — APPROVED** [review]
Thanks for taking this on!

maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**